### PR TITLE
ENH: toggle TORCH_WARN_ONCE to TORCH_WARN for tests

### PR DESCRIPTION
**Pull request by mattip**
Toward fixing #47624

~Step 1: add `TORCH_WARN_MAYBE` which can either warn once or every time in c++, and add a c++ function to toggle the value.
Step 2 will be to expose this to python for tests. Should I continue in this PR or should we take a different approach: add the python level exposure without changing any c++ code and then over a series of PRs change each call site to use the new macro and change the tests to make sure it is being checked?~

Step 1: add a python and c++ toggle to convert TORCH_WARN_ONCE into TORCH_WARN so the warnings can be caught in tests
Step 2: add a python-level decorator to use this toggle in tests
Step 3: (in future PRs): use the decorator to catch the warnings instead of `maybeWarnsRegex`

### CODE REVIEW

File: torch/csrc/Module.cpp
Comment (albanD) [inline torch/csrc/Module.cpp:457]: nit: This usually enclose the whole body of the function (including arg checking and return statement).

**Review by albanD — COMMENTED** [review]
Hi,

While it is clear to me what the difference between `TORCH_WARN` and `TORCH_WARN_ONCE` is. And in which case we should use them.
I am not sure what `TORCH_WARN_MAYBE` means. In particular, does it mean that my warning might get randomly ignored at runtime?

Another approach might be to change the behavior of `TORCH_WARN_ONCE` directly if it is just used for testing?
Also I don't think this needs to be something that can be toggled on/off and maybe a constant flag controlled via an env variable on startup would be enough and avoid issues with mutable global states.

**Comment by mattip** [comment]
@albanD I chose a bad name for the new functionality. Here is what I understand: 

The current mechanism uses `HANDLE_TH_ERRORS` at the entrance to each Python function call to set a c++ warning handler. I understood the point of the issue is, when `HANDLE_TH_ERRORS` is called, to convert `TORCH_WARN_ONCE` into a different creature: one that would warn always and that those warnings would be caught by python. So what would be a better name for this strategy?

**Comment by albanD** [comment]
The `HANDLE_TH_ERRORS` construct is going to convert everything that is being "raised" from c++ into the python equivalent (errors and warnings).
The main different of the `TORCH_WARN_ONCE` version is that it does some dark magic to make sure that the given warning can only be triggered once per process and then will just be ignored.

You want to make `TORCH_WARN_ONCE` always warn but only report it to python the first time?
But then it wouldn't have the right behavior anymore when working with libtorch?

**Comment by mattip** [comment]
@mruberry  request that kicked this off is (emphasis added)
> This feature request is for a mechanism to consistently trigger warnings that use `TORCH_WARN_ONCE`. One mechanism for this would be to add a new context that, if set, **causes these warnings to always warn**. This would let us reliably test that these warnings are thrown correctly.

File: aten/src/ATen/Context.cpp
Comment (mruberry) [inline aten/src/ATen/Context.cpp:68]: I don't think we want to add TORCH_WARN_MAYBE. See my note here:

https://github.com/pytorch/pytorch/issues/47624#issuecomment-731103415

I think we just want to keep TORCH_WARN_ONCE and TORCH_WARN. Yes, TORCH_WARN_ONCE may be a misnomer if this flag is set, but I think that's OK.

File: c10/util/Exception.cpp
Comment (mruberry) [inline c10/util/Exception.cpp:122]: Putting this in context.h and using the same pattern that @kurtamohler did for `_deterministic_ should be a good model: 

https://github.com/pytorch/pytorch/blob/9417e92722f1b6f75e162cbe661d3ed788d8d37a/aten/src/ATen/Context.h#L142

File: c10/util/Exception.h
Comment (mruberry) [inline c10/util/Exception.h:419]: I think we should just replace TORCH_WARN_ONCE with this.

File: torch/csrc/Module.cpp
Comment (mruberry) [reply torch/csrc/Module.cpp:457]: I'm also not sure why this would be added in this PR.

File: c10/util/Exception.cpp
Comment (mruberry) [reply c10/util/Exception.cpp:122]: Like `_determinstic` this also needs to expose the toggling of the variable to Python so we can call it from the Python test suite.

**Review by mruberry — CHANGES_REQUESTED** [review]
Hey @mattip! Thanks for pursuing this problem. It will be nice to have a test suite that can validate warnings are thrown as expected. I made a few comments about the pattern we're going for, I also want to follow-up on @albanD's idea to expose this as a build flag vs a global context.

**Comment by mruberry** [comment]
> While it is clear to me what the difference between `TORCH_WARN` and `TORCH_WARN_ONCE` is. And in which case we should use them.

TORCH_WARN_ONCE warns once per process, TORCH_WARN warns every time an event occurs. I wrote a section in the Developer FAQ about when to use one vs. the other: https://github.com/pytorch/pytorch/wiki/Developer-FAQ. 

> I am not sure what `TORCH_WARN_MAYBE` means. In particular, does it mean that my warning might get randomly ignored at runtime?

I agree. See my comments on the original issue, too. I think we can stick with TORCH_WARN_ONCE and TORCH_WARN. 
 
> Also I don't think this needs to be something that can be toggled on/off and maybe a constant flag controlled via an env variable on startup would be enough and avoid issues with mutable global states.

This is interesting idea, but let me explain the motivation for this flag.

Assume we want to use TORCH_WARN_ONCE, but we also want to verify the warning is thrown as expected. Today we don't have a great way to do this (short of someone opening a subprocess, I guess, which seems like overkill). This is because the warning may have already occurred in the test suite and so, since it's TORCH_WARN_ONCE, won't occur again in the test we write for it. In fact, we even have this:

https://github.com/pytorch/pytorch/blob/85121a7a0f24faccdb36093e92262c186fc65ee0/torch/testing/_internal/common_utils.py#L1239

which actually will warn you if any OTHER warning was raised. But whether the expected warning is raised or not doesn't matter.

The idea behind this request was for a flag where the Python test suite could toggle TORCH_WARN_ONCE into TORCH_WARN to reliably verify these warnings are thrown when they should be. This is still imperfect, unfortunately, because we won't have a great way to verify they're thrown once and only once, but that should be easy to see from code inspection.

Back to your suggestion. Can we make this a build flag? I think the answer is "maybe?" We could make it a build flag and have the relevant tests query for it to know if they should run expecting the warning or not. If the tests are written to assume the flag is always set, however, then the tests won't pass on most PyTorch builds.

Does that make sense, @albanD?

**Comment by albanD** [comment]
> Back to your suggestion. Can we make this a build flag?

Ho sorry for the misunderstanding! I wasn't suggesting a build flag but a runtime flag like [this one](https://github.com/pytorch/pytorch/blob/be45c3401af8186f97f0e2b269ff3bafaf16157f/torch/csrc/Exceptions.cpp#L24-L37) that is read once at startup.
Maybe this is not the best design, but, in my mind at least, it makes the state handling very easy as nothing is mutable (and no need to add a C++ and python API to toggle it). Finally you just need to set an env variable in the `run_test.sh` to enable that for our test suite.

I am not super attached to that idea though. So a mutable global flag is fine here as well and my main concern was about the name, but I think we agree on this one.

**Comment by mattip** [comment]
> So a mutable global flag is fine here as well and my main concern was about the name, but I think we agree on this one.

I will redo the PR along these lines.

**Comment by mruberry** [comment]
> Ho sorry for the misunderstanding! I wasn't suggesting a build flag but a runtime flag like [this one](https://github.com/pytorch/pytorch/blob/be45c3401af8186f97f0e2b269ff3bafaf16157f/torch/csrc/Exceptions.cpp#L24-L37) that is read once at startup.
> Maybe this is not the best design, but, in my mind at least, it makes the state handling very easy as nothing is mutable (and no need to add a C++ and python API to toggle it). Finally you just need to set an env variable in the `run_test.sh` to enable that for our test suite.

Setting an environment variable like this would be OK but we'd have to make sure every test set that needed it set it. A toggle with a context manager might be easier for tests to manage. I could also imagine the niche scenario of wanting to detect warnings only in part of a program.

File: torch/csrc/Module.cpp
Comment (mattip) [reply torch/csrc/Module.cpp]: removed

File: aten/src/ATen/Context.cpp
Comment (mattip) [reply aten/src/ATen/Context.cpp]: reverted

File: c10/util/Exception.cpp
Comment (mattip) [reply c10/util/Exception.cpp:130]: It cannot be in `aten/src/ATEN/Context.h` since the macro is defined in `c10/util/Exception.h` and that is included in `Context.h`, so the value that is toggled must be included before the definition of the macro. I left it in `c10::Warnings`.

File: c10/util/Exception.h
Comment (mattip) [reply c10/util/Exception.h]: done

**Review by mattip — COMMENTED** [review]
- The value can be toggled from python by `torch.set_warnOnlyOnce()` and examined via `torch.is_warnOnlyOnce()`
- I added a `TestCase.mustWarnRegex` decorator that will toggle `TORCH_WARN_ONCE` to `TORCH_WARN` and the check that the desired warning + regex message are emitted. This can be used to replace the `maybeWarnRegex` decorator in subsequent PRs.
- I added a `test_warnOnce_caught` test to `test_torch` and made sure `TORCH_WARN_ONCE` will always warn under the decorator

**Comment by mruberry** [comment]
> * The value can be toggled from python by `torch.set_warnOnlyOnce()` and examined via `torch.is_warnOnlyOnce()`
> * I added a `TestCase.mustWarnRegex` decorator that will toggle `TORCH_WARN_ONCE` to `TORCH_WARN` and the check that the desired warning + regex message are emitted. This can be used to replace the `maybeWarnRegex` decorator in subsequent PRs.
> * I added a `test_warnOnce_caught` test to `test_torch` and made sure `TORCH_WARN_ONCE` will always warn under the decorator

Cool! - is this ready for another review?

**Comment by mattip** [comment]
> Cool! - is this ready for another review?

Sure. I _think_ CI is ok, although it has not finished

File: c10/util/Exception.cpp
Comment (mruberry) [reply c10/util/Exception.cpp:130]: Sounds good.

File: c10/util/Exception.cpp
Comment (mruberry) [inline c10/util/Exception.cpp:133]: What's `ret` for?

File: docs/source/torch.rst
Comment (mruberry) [inline docs/source/torch.rst:558]: cc @kurtamohler for these names, since we just updated the deterministic flags (see https://github.com/pytorch/pytorch/pull/49904)

File: torch/csrc/Module.cpp
Comment (mruberry) [inline torch/csrc/Module.cpp:485]: The internal linter will want this to use Egyptian-style curly braces:

```
if ... {
  Py_RETURN_TRUE;
}
```

File: c10/util/Exception.cpp
Comment (mattip) [reply c10/util/Exception.cpp:133]: leftover from a version that returned the previous value. I will remove in the next iteration.

File: torch/testing/_internal/common_utils.py
Comment (mruberry) [inline torch/testing/_internal/common_utils.py:1306]: I want to suggest a different name for this context manager but I'm struggling to think of one.

Let me try to breakdown how this works so I'm sure I understand it:

- if no warning is thrown it fails the test
- if the expected warning is thrown this context manager exits without issue
- if multiple are warnings then ... they ALL must be the given warning or else this fails an `assertTrue`?

Should this context manager instead verify that the warning it's looking for is thrown, but ignore all other warnings? Or should there be a mechanism to specify that multiple warnings can be thrown and that's OK?

File: test/test_torch.py
Comment (mruberry) [inline test/test_torch.py:3279]: Add the `@onlyCPU` decorator since this test doesn't do anything with the device

File: torch/__init__.py
Comment (mruberry) [inline torch/__init__.py:426]: I think there's an extra underscore here preventing the link from working. See here:

https://9986620-65600975-gh.circle-artifacts.com/0/docs/generated/torch.is_warnOnlyOnce.html

File: torch/__init__.py
Comment (mruberry) [inline torch/__init__.py:415]: Let's write this documentation to be user-facing, so no mention of the macro.

When this flag is True (default) then some PyTorch warnings may only appear once per process. This helps avoid excessive warning information. Setting it to False causes these warnings to always appear, which may be helpful when debugging.

**Comment by mruberry** [comment]
Tests are failing because an override is needed:

```
Jan 06 15:58:59 AssertionError: The following functions are not tested for __torch_function__ support, please ensure there is an entry in the dict returned by torch._overrides.get_testing_overrides for this function or if a __torch_function__ override does not make sense, add an entry to the tuple returned by torch._overrides.get_ignored_functions.
Jan 06 15:58:59 
Jan 06 15:58:59 ["<module 'torch' from "
Jan 06 15:58:59  "'/opt/conda/lib/python3.6/site-packages/torch/__init__.py'>.set_warnOnlyOnce",
Jan 06 15:58:59  "<module 'torch' from "
Jan 06 15:58:59  "'/opt/conda/lib/python3.6/site-packages/torch/__init__.py'>.is_warnOnlyOnce"]
Jan 06 15:58:59 Traceback (most recent call last):
Jan 06 15:58:59   File "test/run_test.py", line 910, in <module>
Jan 06 15:58:59     main()
Jan 06 15:58:59   File "test/run_test.py", line 889, in main
Jan 06 15:58:59     raise RuntimeError(err_message)
Jan 06 15:58:59 RuntimeError: test_overrides failed!
```

File: c10/util/Exception.h
Comment (mruberry) [inline c10/util/Exception.h:160]: Overall the structure of this PR and the test look good. I have some questions about the context manager for testing and suggestions for the documentation, but I'm actually most concerned about the naming.

In particular, "set_warn_once" is ambiguous in its scope. That is: does it set all warnings to warn once? Maybe the flag can be named something like "always_warn", and then these functions would be named something like "set_always_warn" and "get_always_warn_enabled"? @kurtamohler, would that be consistent with our deterministic flag naming thinking?

Then this function would have the conceptual property: when always warn is False (default), some warnings will appear once and then be suppressed for the rest of the process; when always warn is True these warnings will not be suppressed. 

What are your thoughts @mattip?

Comment (mattip) [reply c10/util/Exception.h:160]: I am not really good with names. Your description makes sense to me.

Comment (mruberry) [reply c10/util/Exception.h:160]: > I am not really good with names.

That makes two of us.

File: docs/source/torch.rst
Comment (kurtamohler) [reply docs/source/torch.rst:558]: For the deterministic flag rename, we currently have `torch.use_deterministic_algorithms` and `torch.is_deterministic_algorithms_enabled` on the Python side, `Context::setDeterministicAlgorithms` and `Context::deterministicAlgorithms` on the C++ side.

I think an improvement might be to change `set_warnOnlyOnce` to `set_warn_only_once`.

If we wanted to be extremely strict about consistency with the deterministic flag, we might call it `use_warn_only_once`, but I don't think that grammar makes sense, so probably not a great idea.

File: c10/util/Exception.h
Comment (kurtamohler) [reply c10/util/Exception.h:160]: I think I like the suggestion of changing `warn_once` to `always_warn`. I agree with the idea that it's a better conceptual model to have the default be False.

As for the consistency with deterministic flag names, `is_always_warn_enabled` is probably more consistent on the Python side. But there is also an argument for pairing a function that starts with `set_` with a function that starts with `get_`.

For the C++ side, I think consistency with other functions in `Exception.h` is more important than consistency with `Context::setDeterministicAlgorithms` and `Context::deterministicAlgorithms` in `Context.h`. As far as I can tell, the C++ side of this PR has the right consistency.

File: docs/source/torch.rst
Comment (mattip) [reply docs/source/torch.rst:558]: So for consistency I should replace all `warnOnlyOnce` with `warn_only_once` on both python and C++ level?

File: torch/testing/_internal/common_utils.py
Comment (mattip) [reply torch/testing/_internal/common_utils.py:1306]: I will change it to `assert len(ws) == 1` instead

File: docs/source/torch.rst
Comment (mattip) [reply docs/source/torch.rst:558]: ahh, I see. Later on the reviewers prefer the c++ names to be as is, and only change the python ones.

File: torch/__init__.py
Comment (mattip) [reply torch/__init__.py]: done

File: torch/__init__.py
Comment (mattip) [reply torch/__init__.py]: fixed

File: c10/util/Exception.cpp
Comment (mattip) [reply c10/util/Exception.cpp]: done

File: test/test_torch.py
Comment (mattip) [reply test/test_torch.py]: done

File: torch/csrc/Module.cpp
Comment (mattip) [reply torch/csrc/Module.cpp]: OK. Also modified a few more in the region of this one that used the same pattern.

**Comment by mattip** [comment]
This is ready for another review, assuming CI passes. Since the last review:
- names of the c++ and python functions were changed
- code cleanups
- changed the decorator to check for exactly one warning, and it must match the pattern

**Comment by mattip** [comment]
rebased off latest viable/stable to clear merge conflicts

File: c10/util/Exception.h
Comment (mruberry) [inline c10/util/Exception.h:158]: This comment needs to be updated to reflect the most recent names

File: test/test_torch.py
Comment (mruberry) [inline test/test_torch.py:3292]: Would you elaborate on what's going on here?

File: torch/testing/_internal/common_utils.py
Comment (mruberry) [reply torch/testing/_internal/common_utils.py:1306]: OK I have a different proposal here now. What if we wrap assertWarnsRegex and always enable this setting when we do it? 

See https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertWarnsRegex

It seems like when testing we do want to validate all warnings?

File: torch/overrides.py
Comment (mruberry) [inline torch/overrides.py:186]: I checked similar flags and think we should go with `torch.is_warn_always_enabled` for the public getter name. Sorry about the name thrashing. That's the last name change, I promise.

**Review by mruberry — COMMENTED** [review]
This looks very good. I inlined a few minor tweaks. One last design question I have is whether you think it makes sense to wrap this around unit test's assertRaisesRegex? Curious to hear your thinking.

File: test/test_torch.py
Comment (mattip) [reply test/test_torch.py:3292]: I will add comments

Comment (mattip) [reply test/test_torch.py:3292]: The idea is to check that 
- two consecutive warnings will both be caught, and that 
- the decorator will raise if more than one warning is raised in a test

**Comment by mattip** [comment]
> wrap this around unit test's assertRaisesRegex

Did you mean `assertWarnsRegex`? Yeah, makes sense, although that seems to only check the first warning and not any subsequent ones, although I may be wrong about that.

File: torch/overrides.py
Comment (mattip) [reply torch/overrides.py:186]: done

File: c10/util/Exception.h
Comment (mattip) [reply c10/util/Exception.h]: done

File: torch/testing/_internal/common_utils.py
Comment (mattip) [reply torch/testing/_internal/common_utils.py:1306]: Sorry, just saw this now.

So override the same name or add a new `assertWarnsAlwaysRegex` ?

Comment (mruberry) [reply torch/testing/_internal/common_utils.py:1306]: I think just override assertWarnsRegex to call the base assertWarnsRegex but with this flag set while that happens. So like:

```
def assertWarnsRegex(...):
  try:
    # set flag
    self.super().assertWarnsRegex(...)
  finally:
    # reset flag
```


An unfortunate side-effect of this change is that the test suite will not be thread safe, but oh well.

Comment (mattip) [reply torch/testing/_internal/common_utils.py:1306]: This doesn't work. It somehow captures the warning in a way that fails the test with "AssertionError: UserWarning not triggered". Is it critical that we re-use the super's class method? Here is what I was using"
```
    def assertWarnsRegex(self, *args, **kwargs):
        """Overrided the super class's assertWarnRegex with one that calls
        set_warn_always() to force TORCH_WARN_ONCE to behave like TORCH_WARN
        """
        prev = torch.is_warn_always_enabled()
        torch.set_warn_always(True)
        try:
            return super().assertWarnsRegex(*args, **kwargs)
        finally:
            torch.set_warn_always(prev)
```

Comment (mruberry) [reply torch/testing/_internal/common_utils.py:1306]: What you've implemented looks correct to me - would you elaborate a little on what you're seeing? Is it that assertWarnsRegex only captures the first warning if multiple are thrown, and that by enabling this flag different warnings are thrown first so these checks are failing?

If so, no we don't have to use super. What you had previously where the warnings are reviewed would be preferable. Let's not worry about the name or design tweaks if that's the case.

Comment (mattip) [reply torch/testing/_internal/common_utils.py:1306]: Here is the complete failure:
```
    @onlyCPU
    def test_warn_always_caught(self, device):
        # Check that we can catch a TORCH_WARN_ONCE warning twice
        # since mustWarnRegex uses set_warn_always(True) which changes
        # TORCH_WARN_ONCE to TORCH_WARN
        a = np.arange(10)
        a.flags.writeable = False
        with self.assertWarnsRegex(UserWarning, '.*non-writeable.*'):
            torch.from_numpy(a)
    
        # OK, got it once, now try again
        with self.assertWarnsRegex(UserWarning, '.*non-writeable.*'):
>           torch.from_numpy(a)
E           AssertionError: UserWarning not triggered

test/test_torch.py:3291: AssertionError
```

I would prefer to not inherit the method since it only seems to complicate things.

Comment (mruberry) [reply torch/testing/_internal/common_utils.py:1306]: That's very odd. OK, let's go with the custom content manager that you had earlier for now.

**Comment by mattip** [comment]
@mruberry have we resolved whether to re-use the `TestCase.assertWarnsRegex` ? Once this is in, we can start refactoring tests to use it.

**Comment by mruberry** [comment]
> @mruberry have we resolved whether to re-use the `TestCase.assertWarnsRegex` ? Once this is in, we can start refactoring tests to use it.

Hey @mattip, thanks for the ping. Yes, sorry if this was unclear: https://github.com/pytorch/pytorch/pull/48560#discussion_r556891179. For simplicity I think we should go with your original approach of implementing a separate context manager that doesn't try to override TestCase.assertWarnsRegex. We can revisit that idea later if we like. We can call it "assertWarnsOnceRegex", which is a misleading name but that's fine, as it makes the context manager's intended use clear. 

Sound good?

**Comment by mattip** [comment]
Switched the context manager name to `assertWarnsOnceRegex`

**Comment by mruberry** [comment]
Ping me when this is ready for a final review / merging, @mattip!

**Comment by mattip** [comment]
@mruberry it should be ready for final review

**Review by mruberry — APPROVED** [review]
Thanks @mattip! It's great to finally have this flag and to close this gap in our test coverage.

**Comment by mattip** [comment]
Cool. Although this is a first step: now we should rewrite tests to use the new context manager and add it where a warning is expected

**Comment by mattip** [comment]
@mruberry just to be clear: I am waiting for this to land to do a follow-on PR to rewrite tests

**Comment by mruberry** [comment]
> @mruberry just to be clear: I am waiting for this to land to do a follow-on PR to rewrite tests

I know, sorry :(

Our infra keeps failing when I try to land this (5+ times now). I'm continuing to fight with it today. It's not wrong with the PR. It's just that our landing system wants to see a series of builds pass, and sometimes those services have issues preventing them from running.

**Comment by mruberry** [comment]
Sorry @mattip, would you rebase this? I've tried to merge it 10+ times and each time our infra has failed somewhere. Once you rebase I'll ninja land it and avoid our infra checks. The failures are unrelated to this PR, it's just had bad luck.

**Comment by mattip** [comment]
@mruberry rebased off latest viable/strict

**Comment by mruberry** [comment]
Test failures appear unrelated.

**Comment by mruberry** [comment]
This finally landed! Thanks for being patient, @mattip!

**Pull request closed without merging**
